scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java

```diff
@@ -442,58 +443,71 @@ public final void processMessage(Record record) throws HandshakeException {
-					} else {
-						// continue with the now fully re-assembled message
-						if (messageToProcess instanceof GenericHandshakeMessage) {
-							HandshakeParameter parameter = session.getParameter();
-							if (parameter == null) {
-								throw new IllegalStateException("handshake parameter are required!");
-							}
-							messageToProcess = ((GenericHandshakeMessage) messageToProcess)
-									.getSpecificHandshakeMessage(parameter);
+		int epoch = session.getReadEpoch();
+		if (epoch != record.getEpoch()) {
+			LOGGER.debug("Discarding HANDSHAKE message with wrong epoch received from peer [{}]:{}{}",
+					record.getPeerAddress(), StringUtil.lineSeparator(), record);
+			throw new IllegalArgumentException("processing record with wrong epoch! " + record.getEpoch() + " expected " + epoch);
+		}
+		try {
+			DTLSMessage messageToProcess = inboundMessageBuffer.getNextMessage(record);
+			while (messageToProcess != null) {
+				if (messageToProcess instanceof FragmentedHandshakeMessage) {
+					messageToProcess = handleFragmentation((FragmentedHandshakeMessage) messageToProcess);
+				}
+				if (messageToProcess == null) {
+					// messageToProcess is fragmented and not all parts have been received yet
+				} else {
+					// continue with the now fully re-assembled message
+					if (messageToProcess instanceof GenericHandshakeMessage) {
+						HandshakeParameter parameter = session.getParameter();
+						if (parameter == null) {
+							throw new IllegalStateException("handshake parameter are required!");
 						}
-						if (messageToProcess.getContentType() == ContentType.HANDSHAKE) {
-							// only cancel on HANDSHAKE messages
-							// the very last flight CCS + FINISH
-							// must be not canceled before the FINISH
-							DTLSFlight flight = pendingFlight.get();
-							if (flight != null) {
-								LOGGER.debug("response for flight {} started", flight.getFlightNumber());
-								flight.setResponseStarted();
-							}
+						messageToProcess = ((GenericHandshakeMessage) messageToProcess)
+								.getSpecificHandshakeMessage(parameter);
+					}
+					if (messageToProcess.getContentType() == ContentType.HANDSHAKE) {
+						// only cancel on HANDSHAKE messages
+						// the very last flight CCS + FINISH
+						// must be not canceled before the FINISH
+						DTLSFlight flight = pendingFlight.get();
+						if (flight != null) {
+							LOGGER.debug("response for flight {} started", flight.getFlightNumber());
+							flight.setResponseStarted();
 						}
-						doProcessMessage(messageToProcess);
 					}
+					doProcessMessage(messageToProcess);
+				}
 
-					// process next expected message (if available yet)
-					messageToProcess = inboundMessageBuffer.getNextMessage();
+				// process next expected message (if available yet)
+				messageToProcess = inboundMessageBuffer.getNextMessage();
+			}
+			session.markRecordAsRead(epoch, record.getSequenceNumber());
+			if (epoch < session.getReadEpoch()) {
```

Comment:
I believe I often take few second to get some of your if condition, just because I would have written it reversely :sweat_smile: 

E.g. here in my mind, this is something like : "after we process the message, Is the session readEpoch greater than before ?" <=> if (session.getReadEpoch() > epoch)

Never mind, it's clearly not a big deal :)